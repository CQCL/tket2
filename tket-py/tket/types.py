from __future__ import annotations

from tket._tket.types import HugrType, TypeBound

__all__ = ["HugrType", "TypeBound", "QB_T", "BOOL_T"]


QB_T = HugrType.qubit()
BOOL_T = HugrType.bool()
