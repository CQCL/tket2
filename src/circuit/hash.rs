//! Circuit hashing.

use core::panic;
use std::hash::{Hash, Hasher};

use fxhash::{FxHashMap, FxHasher64};
use hugr::hugr::views::HierarchyView;
use hugr::ops::{LeafOp, OpName, OpTag, OpTrait, OpType};
use hugr::types::TypeBound;
use hugr::{HugrView, Node, Port};
use petgraph::visit::{self as pg, Walker};

use super::Circuit;

/// Circuit hashing utilities.
pub trait CircuitHash<'circ>: HugrView {
    /// Compute hash of a circuit.
    ///
    /// We compute a hash for each command from its operation and the hash of
    /// its predecessors. The hash of the circuit corresponds to the hash of its
    /// output node.
    ///
    /// This hash is independent from the operation traversal order.
    ///
    /// Adapted from Quartz (Apache 2.0)
    /// <https://github.com/quantum-compiler/quartz/blob/2e13eb7ffb3c5c5fe96cf5b4246f4fd7512e111e/src/quartz/tasograph/tasograph.cpp#L410>
    fn circuit_hash(&'circ self) -> u64;
}

impl<'circ, T> CircuitHash<'circ> for T
where
    T: HugrView + HierarchyView<'circ>,
    for<'a> &'a T:
        pg::GraphBase<NodeId = Node> + pg::IntoNeighborsDirected + pg::IntoNodeIdentifiers,
{
    fn circuit_hash(&'circ self) -> u64 {
        let mut hash_state = HashState::default();

        for node in pg::Topo::new(self).iter(self).filter(|&n| n != self.root()) {
            let hash = hash_node(self, node, &mut hash_state);
            hash_state.set_node(self, node, hash);
        }

        hash_state.get_nonlinear(self.output())
    }
}

/// Auxiliary data for circuit hashing.
///
/// Contains previously computed hashes.
#[derive(Clone, Default, Debug)]
struct HashState {
    /// Computed node hashes for each linear output.
    ///
    /// These are removed from the map when consumed.
    pub linear_hashes: FxHashMap<(Node, Port), u64>,
    /// Computed node hashes.
    ///
    /// Only store hashes for nodes with at least one non-linear output.
    pub nonlinear_hashes: FxHashMap<Node, u64>,
}

impl HashState {
    /// Return the hash for a port.
    ///
    /// If the port was linear, it is removed from the map.
    fn take(&mut self, node: Node, port: Port) -> u64 {
        if let Some(hash) = self.linear_hashes.remove(&(node, port)) {
            return hash;
        }
        self.get_nonlinear(node)
    }

    /// Return the hash for a node.
    fn get_nonlinear(&self, node: Node) -> u64 {
        *self.nonlinear_hashes.get(&node).unwrap()
    }

    /// Register the hash for a node.
    fn set_node(&mut self, circ: &impl HugrView, node: Node, hash: u64) {
        let optype = circ.get_optype(node);
        let signature = optype.signature();
        let mut any_nonlinear = false;
        for (port_type, port) in signature
            .output_types()
            .iter()
            .zip(signature.output_ports())
        {
            match TypeBound::Copyable.contains(port_type.least_upper_bound()) {
                true => any_nonlinear = true,
                false => {
                    self.linear_hashes.insert((node, port), hash);
                }
            }
        }
        if any_nonlinear || optype.tag() <= OpTag::Output {
            self.nonlinear_hashes.insert(node, hash);
        }
    }
}

/// Returns a hashable representation of an operation.
///
/// Panics if the operation is a parametric CustomOp
//
// TODO: Hash custom op parameters. Also the extension name?
fn hashable_op(op: &OpType) -> impl Hash {
    match op {
        OpType::LeafOp(LeafOp::CustomOp(op)) if !op.args().is_empty() => {
            panic!("Parametric operation {} cannot be hashed.", op.name())
        }
        _ => op.name(),
    }
}

/// Compute the hash of a circuit command.
///
/// Uses the hash of the operation and the node hash of its predecessors.
///
/// # Panics
/// - If the command is a container node, or if it is a parametric CustomOp.
/// - If the hash of any of its predecessors has not been set.
fn hash_node(circ: &impl HugrView, node: Node, state: &mut HashState) -> u64 {
    let op = circ.get_optype(node);
    let mut hasher = FxHasher64::default();

    if circ.children(node).count() > 0 {
        panic!("Cannot hash container node {node:?}.");
    }
    hashable_op(op).hash(&mut hasher);

    // Add each each input neighbour hash, including the connected ports.
    // TODO: Ignore state edges?
    for input in circ.node_inputs(node) {
        // Combine the hash for each subport, ignoring their order.
        let input_hash = circ
            .linked_ports(node, input)
            .map(|(pred_node, pred_port)| {
                let pred_node_hash = state.take(pred_node, pred_port);
                fxhash::hash64(&(pred_node_hash, pred_port, input))
            })
            .fold(0, |total, hash| hash ^ total);
        input_hash.hash(&mut hasher);
    }
    hasher.finish()
}

#[cfg(test)]
mod test {
    use hugr::hugr::views::{HierarchyView, SiblingGraph};
    use hugr::HugrView;

    use crate::utils::build_simple_circuit;
    use crate::T2Op;

    use super::*;

    #[test]
    fn hash_equality() {
        let hugr1 = build_simple_circuit(2, |circ| {
            circ.append(T2Op::H, [0])?;
            circ.append(T2Op::T, [1])?;
            circ.append(T2Op::CX, [0, 1])?;
            Ok(())
        })
        .unwrap();
        let circ1: SiblingGraph<'_> = SiblingGraph::new(&hugr1, hugr1.root());
        let hash1 = circ1.circuit_hash();

        // A circuit built in a different order should have the same hash
        let hugr2 = build_simple_circuit(2, |circ| {
            circ.append(T2Op::T, [1])?;
            circ.append(T2Op::H, [0])?;
            circ.append(T2Op::CX, [0, 1])?;
            Ok(())
        })
        .unwrap();
        let circ2: SiblingGraph<'_> = SiblingGraph::new(&hugr2, hugr2.root());
        let hash2 = circ2.circuit_hash();

        assert_eq!(hash1, hash2);

        // Inverting the CX control and target should produce a different hash
        let hugr3 = build_simple_circuit(2, |circ| {
            circ.append(T2Op::T, [1])?;
            circ.append(T2Op::H, [0])?;
            circ.append(T2Op::CX, [1, 0])?;
            Ok(())
        })
        .unwrap();
        let circ3: SiblingGraph<'_> = SiblingGraph::new(&hugr3, hugr3.root());
        let hash3 = circ3.circuit_hash();

        assert_ne!(hash1, hash3);
    }
}
