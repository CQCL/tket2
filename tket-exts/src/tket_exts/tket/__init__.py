"""HUGR extension definitions for tket."""

from . import bool, debug, guppy, futures, qsystem, quantum, result, rotation, wasm

__all__ = [
    "bool",
    "debug",
    "gpu",
    "guppy",
    "futures",
    "qsystem",
    "quantum",
    "result",
    "rotation",
    "wasm",
]
