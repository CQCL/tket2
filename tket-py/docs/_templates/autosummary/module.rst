..
   Custom module template to make sphinx-autosummary list the full API doc after
   the summary. See https://github.com/sphinx-doc/sphinx/issues/7912

{{ name | escape | underline}}

.. automodule:: {{ fullname }}

   {% block classes %}
   {% if classes %}
   .. rubric:: {{ _('Classes') }}

   .. autosummary::
      :toctree:
   {% for item in classes %}
      {{ item }}
   {%- endfor %}
   {% endif %}
   {% endblock %}

   {% block functions %}
   {% if functions %}

   .. rubric:: {{ _('Functions') }}

   .. autosummary::
      :toctree:
   {% for item in functions %}
      {{ item }}
   {%- endfor %}
   {% endif %}
   {% endblock %}

   {% block exceptions %}
   {% if exceptions %}
   .. rubric:: {{ _('Exceptions') }}

   .. autosummary::
      :toctree:
   {% for item in exceptions %}
      {{ item }}
   {%- endfor %}
   {% endif %}
   {% endblock %}

{% block modules %}
{% if modules %}
.. autosummary::
   :toctree:
   :recursive:
{% for item in modules %}
   {{ item }}
{%- endfor %}
{% endif %}
{% endblock %}