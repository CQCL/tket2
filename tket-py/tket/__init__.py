"""TKET is an open source quantum compiler developed by Quantinuum.

Central to TKET's design is its hardware agnosticism which allows researchers
and quantum software developers to take advantage of its powerful compilation
for many different quantum architectures.

This is the Python front-end for TKET, providing a high-level interface for
working with quantum circuits. See also the Rust library with the same name on
[crates.io](https://crates.io/crates/tket).
"""

from . import circuit, ops, optimiser, passes, pattern, rewrite, matcher, protocol

__all__ = [
    "circuit",
    "matcher",
    "ops",
    "optimiser",
    "passes",
    "pattern",
    "rewrite",
    "protocol",
]


# This is updated by our release-please workflow, triggered by this
# annotation: x-release-please-version
__version__ = "0.12.7"
