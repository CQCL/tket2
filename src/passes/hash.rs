//use super::{pattern::node_equality, CircFixedStructPattern};
use crate::circuit::{circuit::Circuit, dag::Dag, operation::Op};
use portgraph::{
    graph::{Direction, NodeIndex},
    toposort::TopSortWalker,
};
use std::collections::{HashMap, VecDeque};

fn op_hash(op: &Op) -> Option<usize> {
    Some(match op {
        Op::H => 1,
        Op::CX => 2,
        Op::ZZMax => 3,
        Op::Reset => 4,
        // These shouldn't happen in the normal course of hashing
        // Op::Input => 5,
        // Op::Output => 6,
        Op::Noop(_) => 7,
        Op::Measure => 8,
        Op::Barrier => 9,
        Op::AngleAdd => 10,
        Op::AngleMul => 11,
        Op::AngleNeg => 12,
        Op::QuatMul => 13,
        // Op::Copy { n_copies, typ } => todo!(),
        Op::RxF64 => 14,
        Op::RzF64 => 15,
        Op::TK1 => 16,
        Op::Rotation => 17,
        Op::ToRotation => 18,
        // should Const of different values be hash different?
        Op::Const(_) => 19,
        // Op::Custom(_) => todo!(),
        _ => return None,
    })
}

#[derive(Debug, Clone)]
pub struct PermHash {
    hash_val: usize,
    output_order: Vec<usize>,
}

fn combine_non_assoc(ph: &PermHash, portnum: usize) -> PermHash {
    PermHash {
        hash_val: ph.hash_val * (portnum + 1),
        output_order: ph.output_order.clone(),
    }
}

fn hash_node(dag: &Dag, n: NodeIndex, edge_hashes: impl IntoIterator<Item = PermHash>) -> PermHash {
    let op_hash =
        op_hash(&dag.node_weight(n).expect("No weight for node").op).expect("Unhashable op");

    let (edge_hashes, edge_outords): (Vec<usize>, Vec<Vec<usize>>) = edge_hashes
        .into_iter()
        .enumerate()
        .map(|(near_portnum, far_end_hash)| combine_non_assoc(&far_end_hash, near_portnum))
        .map(|ph| (ph.hash_val, ph.output_order))
        .unzip();
    PermHash {
        hash_val: edge_hashes
            .iter()
            // Edge combining function here. Of course the arithmetic gets chained, so for three edges we'll have 3(3a + 5b) + 5c == 9a + 15b + 5c
            .fold(op_hash, |nh, eh| (3 * nh) + (5 * eh)),
        output_order: edge_outords.into_iter().flatten().collect(),
    }
}

// TODO take an iterator of edge classes rather than just a number
fn hash_ports(node_hash: PermHash, edges: usize) -> Vec<PermHash> {
    (0..edges)
        .map(|i| combine_non_assoc(&node_hash, i))
        .collect()
}

pub fn invariant_hash(circ: &Circuit) -> Vec<PermHash> {
    // Firstly compute "forwards" (depending on inputs) hashes of parts
    // of the graph that do not depend upon the graph input.
    let mut fwd_hashes: HashMap<NodeIndex, Vec<PermHash>> = HashMap::new();
    let source_nodes = circ
        .dag
        .node_indices()
        .filter(|n| circ.dag.node_edges(*n, Direction::Incoming).all(|_| false))
        .collect();
    for n in TopSortWalker::new(&circ.dag, source_nodes) {
        // Note, could make this more efficient using a worklist rather than topsort,
        // as topsort continues traversal over the entire graph whereas we could tell when to stop.
        if circ.dag.node_weight(n).unwrap().op == Op::Input {
            continue;
        }
        if !circ.dag.node_edges(n, Direction::Incoming).all(|e| {
            fwd_hashes.contains_key(&circ.dag.edge_endpoint(e, Direction::Incoming).unwrap())
        }) {
            continue;
        }
        let node_hash = hash_node(
            &circ.dag,
            n,
            circ.dag.node_edges(n, Direction::Incoming).map(|e| {
                let in_node = circ.dag.edge_endpoint(e, Direction::Incoming).unwrap();
                fwd_hashes.get(&in_node).unwrap()
                    [circ.port_of_edge(in_node, e, Direction::Outgoing).unwrap()]
                .clone()
            }),
        );
        fwd_hashes.insert(
            n,
            hash_ports(
                node_hash,
                circ.dag.node_edges(n, Direction::Outgoing).count(),
            ),
        );
    }
    // Now hash rest of circuit backwards, from output back to input
    let output_nodes: VecDeque<NodeIndex> = circ
        .dag
        .node_indices()
        .filter(|n| circ.dag.node_edges(*n, Direction::Outgoing).all(|_| false))
        .collect();
    assert!(
        output_nodes.len() == 1
            && circ.dag.node_weight(output_nodes[0]).map(|v| v.op.clone()) == Some(Op::Output)
    );
    // hash per input for nodes with inputs
    let mut node_hashes: HashMap<NodeIndex, Vec<PermHash>> = HashMap::new();
    let mut input_hash = None;
    node_hashes.insert(
        output_nodes[0],
        circ.dag
            .node_edges(output_nodes[0], Direction::Incoming)
            .enumerate()
            .map(|(i, _)| PermHash {
                hash_val: 6,
                output_order: [i].to_vec(),
            })
            .collect(),
    );
    for n in TopSortWalker::new(&circ.dag, output_nodes).reversed() {
        // Again, could possibly make this a bit more efficient using a worklist,
        // but the major part of that saving would probably be achieved by returning early
        if fwd_hashes.contains_key(&n) {
            continue;
        }
        let edge_targets = circ.dag.node_edges(n, Direction::Outgoing).map(|e| {
            let target = circ.dag.edge_endpoint(e, Direction::Incoming).unwrap();
            let target_port = circ.port_of_edge(target, e, Direction::Incoming).unwrap();
            node_hashes.get(&target).expect("Edge target not found")[target_port].clone()
            // TODO where output ports are equivalent, use same source_port
        });
        if circ.dag.node_weight(n).unwrap().op == Op::Input {
            // we could just return here, but we'll continue as a sanity check
            assert!(input_hash.is_none());
            input_hash = Some(edge_targets.collect());
        } else {
            // Also include fwd_hashes of any *incoming* edge
            let in_edges = circ.dag.node_edges(n, Direction::Incoming).flat_map(|e| {
                let src_node = circ.dag.edge_endpoint(e, Direction::Outgoing).unwrap();
                fwd_hashes.get(&src_node).map(|v| {
                    v[circ.port_of_edge(src_node, e, Direction::Outgoing).unwrap()].clone()
                })
            });
            let node_hash = hash_node(&circ.dag, n, edge_targets.chain(in_edges));

            node_hashes.insert(
                n,
                hash_ports(
                    node_hash,
                    circ.dag.node_edges(n, Direction::Incoming).count(),
                ),
            );
        }
    }
    input_hash.unwrap()
}

pub fn circuit_hash(circ: &Circuit) -> usize {
    let mut total: usize = 0;

    let mut hash_vals: HashMap<NodeIndex, usize> = HashMap::new();
    let [i, _] = circ.boundary();

    let _ophash = |o| 17 * 13 + op_hash(o).expect("unhashable op");
    hash_vals.insert(i, _ophash(&Op::Input));

    let initial_nodes = circ
        .dag
        .node_indices()
        .filter(|n| circ.dag.node_edges(*n, Direction::Incoming).count() == 0)
        .collect();

    for nid in TopSortWalker::new(circ.dag_ref(), initial_nodes) {
        if hash_vals.contains_key(&nid) {
            continue;
        }

        let mut myhash = _ophash(circ.node_op(nid).expect("op not found."));

        for ine in circ.node_edges(nid, Direction::Incoming) {
            let (src, _) = circ.edge_endpoints(ine).expect("edge not found.");
            debug_assert!(hash_vals.contains_key(&src));

            let mut edgehash = hash_vals[&src];

            // TODO check if overflow arithmetic is intended

            edgehash = edgehash.wrapping_mul(31).wrapping_add(
                circ.port_of_edge(src, ine, Direction::Outgoing)
                    .expect("edge not found."),
            );
            edgehash = edgehash.wrapping_mul(31).wrapping_add(
                circ.port_of_edge(nid, ine, Direction::Incoming)
                    .expect("edge not found."),
            );

            myhash = myhash.wrapping_add(edgehash);
        }
        hash_vals.insert(nid, myhash);
        total = total.wrapping_add(myhash);
    }

    total
}

#[cfg(test)]
mod tests {
    use crate::circuit::{
        circuit::UnitID,
        operation::{ConstValue, WireType},
    };

    use super::*;
    #[test]
    fn test_simple_hash() {
        let mut circ1 = Circuit::new();
        let [input, output] = circ1.boundary();

        let point5 = circ1.add_vertex(Op::Const(ConstValue::f64_angle(0.5)));
        let rx = circ1.add_vertex(Op::RzF64);
        circ1
            .add_insert_edge((input, 0), (rx, 0), WireType::Qubit)
            .unwrap();
        circ1
            .add_insert_edge((point5, 0), (rx, 1), WireType::Angle)
            .unwrap();
        circ1
            .add_insert_edge((rx, 0), (output, 0), WireType::Qubit)
            .unwrap();

        let mut circ = Circuit::new();
        let [input, output] = circ.boundary();

        // permute addition operations
        let rx = circ.add_vertex(Op::RzF64);
        let point5 = circ.add_vertex(Op::Const(ConstValue::f64_angle(0.5)));
        circ.add_insert_edge((rx, 0), (output, 0), WireType::Qubit)
            .unwrap();
        circ.add_insert_edge((input, 0), (rx, 0), WireType::Qubit)
            .unwrap();
        circ.add_insert_edge((point5, 0), (rx, 1), WireType::Angle)
            .unwrap();

        assert_eq!(circuit_hash(&circ), circuit_hash(&circ1));

        let mut circ = Circuit::new();
        let [input, output] = circ.boundary();

        let point5 = circ.add_vertex(Op::Const(ConstValue::f64_angle(0.5)));
        let rx = circ.add_vertex(Op::RxF64); // rx rather than rz
        circ.add_insert_edge((input, 0), (rx, 0), WireType::Qubit)
            .unwrap();
        circ.add_insert_edge((point5, 0), (rx, 1), WireType::Angle)
            .unwrap();
        circ.add_insert_edge((rx, 0), (output, 0), WireType::Qubit)
            .unwrap();

        assert_ne!(circuit_hash(&circ), circuit_hash(&circ1));
    }
}
