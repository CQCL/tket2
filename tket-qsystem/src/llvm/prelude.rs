//! LLVM lowering implementations for HUGR prelude operations.

use std::hash::{DefaultHasher, Hash as _, Hasher as _};
use tket::hugr::{self, llvm::inkwell};

use anyhow::{anyhow, ensure, Result};
use hugr::llvm::emit::func::EmitFuncContext;
use hugr::llvm::extension::prelude::PreludeCodegen;
use hugr::llvm::types::TypingSession;
use inkwell::attributes::{Attribute, AttributeLoc};
use inkwell::module::Linkage;
use inkwell::types::{BasicType, IntType};
use inkwell::values::{BasicValue, BasicValueEnum, FunctionValue, StructValue};
use inkwell::AddressSpace;
use tket::hugr::{HugrView, Node};

#[derive(Clone)]
/// Codegen extension for the QIS prelude.
pub struct QISPreludeCodegen;

impl PreludeCodegen for QISPreludeCodegen {
    fn usize_type<'c>(&self, session: &TypingSession<'c, '_>) -> IntType<'c> {
        session.iw_context().i64_type()
    }

    /// Return the llvm type of [`tket::hugr::extension::prelude::qb_t`].
    fn qubit_type<'c>(&self, session: &TypingSession<'c, '_>) -> impl BasicType<'c> {
        session.iw_context().i64_type()
    }

    /// A panic ends all shots, this is done by using a return code > 1000.
    /// The lowering of the panic HUGR op shifts the return code held by the op by 1000.
    /// So valid values for `signal` in the panic HUGR op are 1-(2^31 - 1001).
    fn emit_panic<H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<H>,
        err: BasicValueEnum,
    ) -> Result<()> {
        let err = StructValue::try_from(err)
            .map_err(|()| anyhow!("emit_panic: Expected err to have struct type"))?;
        let return_code = ctx.builder().build_extract_value(err, 0, "")?;
        // TODO check if code is in valid range 1-(2^31 - 1001), otherwise error.
        // https://github.com/quantinuum-dev/eldarion/issues/287
        let return_code = {
            let shift = ctx.iw_context().i32_type().const_int(1000, false);
            ctx.builder()
                .build_int_add(return_code.into_int_value(), shift, "shift_code")?
        };

        let msg = ctx.builder().build_extract_value(err, 1, "")?;
        let panic = Self::get_panic(ctx)?;
        ctx.builder()
            .build_call(panic, &[return_code.into(), msg.into()], "")?;
        Ok(())
    }

    /// An exit ends the current shot using a return code `rc`, where `1 <= rc <= 1000`.
    fn emit_exit<H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<H>,
        err: BasicValueEnum,
    ) -> Result<()> {
        let err = StructValue::try_from(err)
            .map_err(|()| anyhow!("emit_exit: Expected err to have struct type"))?;
        let return_code = ctx.builder().build_extract_value(err, 0, "")?;
        // TODO check if code is in valid range 1-1000, otherwise error.
        // https://github.com/quantinuum-dev/eldarion/issues/287
        let msg = ctx.builder().build_extract_value(err, 1, "")?;
        let panic = Self::get_panic(ctx)?;
        ctx.builder()
            .build_call(panic, &[return_code.into(), msg.into()], "")?;
        Ok(())
    }

    fn emit_print<H: HugrView<Node = Node>>(
        &self,
        _ctx: &mut EmitFuncContext<H>,
        _text: inkwell::values::BasicValueEnum,
    ) -> Result<()> {
        Ok(())
    }

    fn emit_const_error<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        err: &hugr::extension::prelude::ConstError,
    ) -> Result<BasicValueEnum<'c>> {
        let err_ty = ctx
            .llvm_type(&hugr::extension::prelude::error_type())?
            .into_struct_type();
        let signal = err_ty
            .get_field_type_at_index(0)
            .unwrap()
            .into_int_type()
            .const_int(u64::from(err.signal), false);

        let message = emit_global_string(ctx, &err.message, "e_", "EXIT:INT:")?;
        let err = err_ty.const_named_struct(&[signal.into(), message]);
        Ok(err.into())
    }

    fn emit_const_string<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        str: &hugr::extension::prelude::ConstString,
    ) -> Result<BasicValueEnum<'c>> {
        emit_global_string(ctx, str.value(), "s_", "")
    }
}

impl QISPreludeCodegen {
    fn get_panic<'c, H: HugrView<Node = Node>>(
        ctx: &EmitFuncContext<'c, '_, H>,
    ) -> Result<FunctionValue<'c>> {
        let iwc = ctx.iw_context();
        let func_type = iwc.void_type().fn_type(
            &[
                iwc.i32_type().into(),
                iwc.i8_type().ptr_type(AddressSpace::default()).into(),
            ],
            false,
        );
        let f = ctx.get_extern_func("panic", func_type)?;
        let noreturn = Attribute::get_named_enum_kind_id("noreturn");
        debug_assert!(noreturn != 0);
        f.add_attribute(
            AttributeLoc::Function,
            iwc.create_enum_attribute(noreturn, 0),
        );

        Ok(f)
    }
}

/// Emit a global string constant with a unique name.
///
/// The symbol name is constructed with `symbol_prefix` and ``str`.
///
/// The string is prefixed with `str_prefix`.
pub fn emit_global_string<'c, H: HugrView<Node = Node>>(
    ctx: &EmitFuncContext<'c, '_, H>,
    str: impl AsRef<str>,
    symbol_prefix: impl AsRef<str>,
    str_prefix: impl AsRef<str>,
) -> Result<BasicValueEnum<'c>> {
    let (str, symbol_prefix, str_prefix) =
        (str.as_ref(), symbol_prefix.as_ref(), str_prefix.as_ref());
    let tagged_str = format!("{str_prefix}{str}");
    let tagged_str_bytes = tagged_str.as_bytes();
    let tagged_str_len = tagged_str_bytes.len();
    ensure!(
        tagged_str_len < 256,
        "Constant string too long: {tagged_str_len} >= 256"
    );

    let tgt_str_bytes = [&[u8::try_from(tagged_str_len)?], tagged_str_bytes].concat();
    let tgt_str_value = ctx.iw_context().const_string(&tgt_str_bytes, false);

    let global = {
        // we need a unique name for the global. We use a prefix(of length
        // 10) of the string, followed by the hash of the whole string.
        //
        // This should be enough, but for absolute guarantee of success we
        // append ".i", i an integer and find an available name.
        let hash = {
            let mut s = DefaultHasher::default();
            tgt_str_bytes.hash(&mut s);
            let h = s.finish();
            let lower32 = h as u32;
            format!("{lower32:08X}")
        };
        // Use unicode-aware character slicing to prevent slicing through multi-byte characters
        let prefix_chars: String = str.chars().take(10).collect();
        let name_prefix = format!("{symbol_prefix}{prefix_chars}.{hash}");

        (0..)
            .find_map(|i| {
                let name = format!("{name_prefix}.{i}");
                if let Some(global) = ctx.get_current_module().get_global(&name) {
                    if let Some(v) = global.get_initializer() {
                        if v == tgt_str_value {
                            // We've found a global with the same value, use that.
                            return Some(global);
                        }
                    }
                } else {
                    let global =
                        ctx.get_current_module()
                            .add_global(tgt_str_value.get_type(), None, &name);
                    global.set_initializer(&tgt_str_value);
                    global.set_linkage(Linkage::Private);
                    global.set_constant(true);
                    return Some(global);
                }
                None
            })
            .unwrap()
    };
    Ok(ctx
        .builder()
        .build_pointer_cast(
            global.as_pointer_value(),
            ctx.iw_context().i8_type().ptr_type(AddressSpace::default()),
            "",
        )?
        .as_basic_value_enum())
}

#[cfg(test)]
mod test {
    use hugr::builder::{Dataflow, DataflowHugr};
    use hugr::extension::prelude::{self, qb_t, ConstError, EXIT_OP_ID, PANIC_OP_ID};
    use hugr::extension::PRELUDE;
    use hugr::llvm::check_emission;
    use hugr::llvm::emit::test::SimpleHugrConfig;

    use hugr::llvm::test::llvm_ctx;
    use hugr::types::TypeArg;
    use rstest::rstest;

    use super::*;

    #[rstest]
    fn test_panic_emit() {
        let mut llvm_ctx = llvm_ctx(0);
        let prelude_codegen = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| ceb.add_prelude_extensions(prelude_codegen.clone()));

        // Create a hugr that has a panic message
        let error_val = ConstError::new(42, "PANIC");
        let type_arg_q = TypeArg::Runtime(qb_t());
        let type_arg_2q = TypeArg::List(vec![type_arg_q.clone(), type_arg_q]);
        let panic_op = PRELUDE
            .instantiate_extension_op(&PANIC_OP_ID, [type_arg_2q.clone(), type_arg_2q.clone()])
            .unwrap();

        let hugr = SimpleHugrConfig::new()
            .with_ins(vec![qb_t(), qb_t()])
            .with_outs(vec![qb_t(), qb_t()])
            .with_extensions(prelude::PRELUDE_REGISTRY.to_owned())
            .finish(|mut builder| {
                let [q0, q1] = builder.input_wires_arr();
                let err = builder.add_load_value(error_val);
                let [q0, q1] = builder
                    .add_dataflow_op(panic_op, [err, q0, q1])
                    .unwrap()
                    .outputs_arr();
                builder.finish_hugr_with_outputs([q0, q1]).unwrap()
            });

        check_emission!(hugr, llvm_ctx);
    }

    #[rstest]
    fn test_exit_emit() {
        let mut llvm_ctx = llvm_ctx(0);
        let prelude_codegen = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| ceb.add_prelude_extensions(prelude_codegen.clone()));

        let error_val = ConstError::new(42, "EXIT");
        let type_arg_q: TypeArg = TypeArg::Runtime(qb_t());
        let type_arg_2q: TypeArg = TypeArg::List(vec![type_arg_q.clone(), type_arg_q]);
        let exit_op = PRELUDE
            .instantiate_extension_op(&EXIT_OP_ID, [type_arg_2q.clone(), type_arg_2q.clone()])
            .unwrap();

        let hugr = SimpleHugrConfig::new()
            .with_ins(vec![qb_t(), qb_t()])
            .with_outs(vec![qb_t(), qb_t()])
            .with_extensions(prelude::PRELUDE_REGISTRY.to_owned())
            .finish(|mut builder| {
                let [q0, q1] = builder.input_wires_arr();
                let err = builder.add_load_value(error_val);
                let [q0, q1] = builder
                    .add_dataflow_op(exit_op, [err, q0, q1])
                    .unwrap()
                    .outputs_arr();
                builder.finish_hugr_with_outputs([q0, q1]).unwrap()
            });

        check_emission!(hugr, llvm_ctx);
    }
}
