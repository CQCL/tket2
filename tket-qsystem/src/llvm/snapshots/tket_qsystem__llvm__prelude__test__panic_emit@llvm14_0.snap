---
source: tket-qsystem/src/llvm/prelude.rs
expression: mod_str
---
; ModuleID = 'test_context'
source_filename = "test_context"

@e_PANIC.DF25FD88.0 = private constant [15 x i8] c"\0EEXIT:INT:PANIC"

define private { i64, i64 } @_hl.main.1(i64 %0, i64 %1) {
alloca_block:
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  %2 = extractvalue { i32, i8* } { i32 42, i8* getelementptr inbounds ([15 x i8], [15 x i8]* @e_PANIC.DF25FD88.0, i32 0, i32 0) }, 0
  %shift_code = add i32 %2, 1000
  %3 = extractvalue { i32, i8* } { i32 42, i8* getelementptr inbounds ([15 x i8], [15 x i8]* @e_PANIC.DF25FD88.0, i32 0, i32 0) }, 1
  call void @panic(i32 %shift_code, i8* %3)
  %mrv = insertvalue { i64, i64 } undef, i64 0, 0
  %mrv8 = insertvalue { i64, i64 } %mrv, i64 0, 1
  ret { i64, i64 } %mrv8
}

; Function Attrs: noreturn
declare void @panic(i32, i8*) #0

attributes #0 = { noreturn }
